runtime/datastructures/src/hook_map.rs

```diff
@@ -0,0 +1,404 @@
+use super::decls::{Map,Set,List,Int,K,__gmpz_set_ui};
+use std::iter::FromIterator;
+use std::ptr;
+use std::mem;
+
+#[no_mangle]
+pub extern "C" fn size_map() -> usize {
+  mem::size_of::<Map>()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn drop_map(ptr: *mut Map) {
+  ptr::drop_in_place(ptr)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_element(result: *mut Map, key: K, value: K) -> bool {
+  ptr::write(result, Map::singleton(key, value));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_unit(result: *mut Map) -> bool {
+  ptr::write(result, Map::new());
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_concat(result: *mut Map, m1: *const Map, m2: *const Map) -> bool {
+  let mut status = true;
+  ptr::write(result, (*m1).clone().union_with((*m2).clone(), |v1, _| { status = false; v1 }));
+  status
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_lookup(result: *mut K, m: *const Map, key: K) -> bool {
+  match (*m).get(&key) {
+    Some(v) => { ptr::write(result, *v); true }
+    None => false
+  }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_lookupOrDefault(result: *mut K, m: *const Map, key: K, default: K) -> bool {
+  let val = (*m).get(&key);
+  match val {
+    Some(v) => *result = *v,
+    None => *result = default
+  }
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_update(result: *mut Map, m: *const Map, key: K, value: K) -> bool {
+  ptr::write(result, (*m).update(key, value));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_remove(result: *mut Map, m: *const Map, key: K) -> bool {
+  ptr::write(result, (*m).without(&key));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_difference(result: *mut Map, m1: *const Map, m2: *const Map) -> bool {
+  ptr::write(result, (*m1).clone().difference_with((*m2).clone(), |v1, v2| if v1 == v2 { None } else { Some(v1) }));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_keys(result: *mut Set, m: *const Map) -> bool {
+  ptr::write(result, Set::from_iter((*m).keys().map(|k| *k)));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_keys_list(result: *mut List, m: *const Map) -> bool {
+  ptr::write(result, List::from_iter((*m).keys().map(|k| *k)));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_in_keys(result: *mut bool, key: K, m: *const Map) -> bool {
+  ptr::write(result, (*m).contains_key(&key));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_values(result: *mut List, m: *const Map) -> bool {
+  ptr::write(result, List::from_iter((*m).values().map(|k| *k)));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_choice(result: *mut K, m: *const Map) -> bool {
+  if (*m).is_empty() {
+    return false;
+  }
+  ptr::write(result, *(*m).keys().next().unwrap());
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_size(result: *mut Int, m: *const Map) -> bool {
+  __gmpz_set_ui(result, (*m).len());
```

Comment:
Does this require the int to already be initialized? Many of the other methods work with uninitialized targets, and would leak memory if called with result pointing to an initialized value.

Reply:
the assumption is that some initialization will have already occurred when the heap block is created, and I'm including the call to mpz_init in that category.